### PR TITLE
#67-2: Remove extra evaluations for bindings by introducing a cached set of bindings

### PR DESCRIPTION
**Pull request by everestbt**
This works by performing the following:
1) Run the securing script as before
2) Pull out the produced bindings and cache them
3) Create a new set of bindings and put those into the engine
4) If custom bindings are present put the cached set of secure bindings into them.

Additional parts done here, if before a script is run it detects that the secure settings have been changed inside the engine then it will re-create them. I am not certain this is needed as this doesn't protect a single running script but it will stop one script from causing problems for another. To address this fully further sanitation of the scripts going in would be needed but that seems outside the scope. 

Typically the cached bindings only contains 16 entries so this check should not cost anything but the reset of the bindings does cost something, however it should rarely take place unless someone makes a mistake.

Important to note this issue discovered for binding still exists for ScriptContext, but at the moment I don't have time to dig into this. I will open an issue to flag this after this pull request and the current issue is closed.

From my testing this resolves issue #67, any comments on the code would be appreciated.

### CODE REVIEW

File: src/main/java/delight/nashornsandbox/internal/NashornSandboxImpl.java [hunk 1]
```diff
@@ -96,40 +101,73 @@ public NashornSandboxImpl(String... params) {
...
-            sb.append("var print=function(){};var echo = function(){};");
+        catch (final Exception e) {
+            throw new RuntimeException(e);
         }
-        if (!allowReadFunctions) {
-            sb.append("var readFully=function(){};").append("var readLine=function(){};");
+    }
+
+    // This will detect whether the current engine bindings match the cached protected bindings
+    private boolean engineBindingsSafe() {
```
Comment (mxro) [inline src/main/java/delight/nashornsandbox/internal/NashornSandboxImpl.java]: Could we rename to `engineBindingsUnchanged`?

Comment (everestbt) [reply src/main/java/delight/nashornsandbox/internal/NashornSandboxImpl.java]: I have renamed it as suggested, it is much clearer this way.

File: src/test/java/delight/nashornsandbox/TestPerformance.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        NashornSandbox sandbox = NashornSandboxes.create();
+
+        String js = "1==1";
+
+
+        long start = System.nanoTime();
+        for (int i = 0; i< 1000; ++i) {
+            sandbox.eval(js);
+        }
+        System.out.println("No binding " + (((double)(System.nanoTime() - start))/1_000_000_000));
```
[the remark this answers is not in the document]
Comment (everestbt) [reply src/test/java/delight/nashornsandbox/TestPerformance.java]: I have opted to remove the test file entirely, it was mostly for my own purposes to see what the difference was between the implementations. I agree they are useless tests, and any asserts would be flaky and prone to failure. They will exist in the history if anyone wants to revive them for some purpose.

File: src/test/java/delight/nashornsandbox/TestPerformance.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import org.junit.Ignore;
+import org.junit.Test;
+
+import javax.script.Bindings;
+import javax.script.ScriptException;
+
+@Ignore
+public class TestPerformance {
+    @Test
+    public void noBindings() throws ScriptCPUAbuseException, ScriptException {
```
[the remark this answers is not in the document]
Comment (everestbt) [reply src/test/java/delight/nashornsandbox/TestPerformance.java]: I have checked the added tests and ensured they follow this pattern.

**Comment by everestbt** [comment]
@mxro Thank you very much for the comments, I have addressed them as best I can. Let me know if you have any other thoughts.

**Comment by mxro** [comment]
@everestbt Perfect, thank you! Sorry again for merging this in later. Merged to master now but not deployed in a release yet.

**Pull request merged**